### PR TITLE
Revised closure caching. Simplified algorithm.

### LINKED ISSUE #4518: Assign extra properties to the object leads to query parsing fail
### Assign extra properties to the object leads to query parsing fail.
```
Exception message:
Constant 'value(xx.xxx+<>c__DisplayClass63_1)' is not replaced.
Stack trace:
```

### Steps to reproduce

```c#
          var extraProperty = new Dictionary<int, int>();

          var test = dc.EntityWithString
            .Select(i => new EntityView { ExtraProperty = extraProperty, String = i.String })
            .Select(i => i.String)
            .Distinct()
            .OrderBy(i => i)
            .ToList();

```

### Environment details

`Linq To DB` version: 6.0.0.0

Database (with version): SQL Server 2017

ADO.NET Provider (with version): Microsoft.Data.SqlClient 3.1.15

Operating system: Windows 10

.NET Version: 4.8

### PR DESCRIPTION
**Pull request by sdanyliv**
Fixes #4518

### CODE REVIEW

File: Source/LinqToDB/Linq/Query.cs [hunk 1]
```diff
@@ -579,72 +563,16 @@ public static Query<T> GetQuery(IDataContext dataContext, ref Expression expr, o
 
 			if (useCache && !query.DoNotCache)
 			{
-				// All non-value type parametrized expression will be transformed to constant. It prevents from caching big reference classes in cache.
+				// All non-value type expression and parametrized expressions will be transformed to ConstantPlaceholderExpression. It prevents from caching big reference classes in cache.
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/Query.cs:566]:
```suggestion
				// All non-value type expressions and parametrized expressions will be transformed to ConstantPlaceholderExpression. It prevents from caching big reference classes in cache.
```

**Review by MaceWindu — APPROVED** [review]

File: Source/Directory.Build.props [hunk 2]
```diff
@@ -12,7 +12,7 @@
 	</PropertyGroup>
 
 	<ItemGroup>
-		<PackageReference Include="Microsoft.CodeAnalysis.PublicApiAnalyzers" />
+		<!-- <PackageReference Include="Microsoft.CodeAnalysis.PublicApiAnalyzers" /> -->
```
Comment (viceroypenguin) [inline Source/Directory.Build.props:15]: disabled reference intended?

File: Source/LinqToDB/Linq/Query.cs [hunk 3]
```diff
@@ -69,10 +69,10 @@ protected bool Compare(IDataContext dataContext, Expression expr)
 				InlineParameters        == dataContext.InlineParameters                                                 &&
 				ContextType             == dataContext.GetType()                                                        &&
 				IsEntityServiceProvided == dataContext is IInterceptable<IEntityServiceInterceptor> { Interceptor: {} } &&
-				Expression!.EqualsTo(expr, dataContext, _parametrized, _parametersDuplicates, _dynamicAccessors);
+				Expression!.EqualsTo(expr, dataContext, _parametersDuplicates, _dynamicAccessors);
 		}
 
-		List<Expression>?                                _parametrized;
+		List<Expression>? _parametrized;
```
Comment (viceroypenguin) [inline Source/LinqToDB/Linq/Query.cs:75]: nit: English spelling is `parameterized`. Dunno if important to change though.

File: Source/Directory.Build.props (hunk 2 above)
Comment (MaceWindu) [reply Source/Directory.Build.props:15]: yep, I will update baselines before release. Those analyzers are PITA at current stage :)

File: Source/LinqToDB/Linq/Query.cs (hunk 3 above)
Comment (viceroypenguin) [reply Source/LinqToDB/Linq/Query.cs:75]: I'll do a project-wide PR for this later

**Review by viceroypenguin — APPROVED** [review]

**Pull request merged**